File: leetcode/328-odd-even-linked-list/main.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

/*
	navie solution
	put the items into even and odd arrays
	concat them together
	time		O(n)
	space	O(n)
	beats 100% (LOL the speed is the same with the classic approach)
*/
func oddEvenList(head *ListNode) *ListNode {
	odd := []*ListNode{}
	even := []*ListNode{}
	cnt := 1
	cur := head
	for cur != nil {
		if cnt%2 == 0 {
			even = append(even, cur)
		} else {
			odd = append(odd, cur)
		}
		cur = cur.Next
		cnt++
	}
	concat := append(odd, even...)

	dump := &ListNode{0, nil}
	cur = dump
	for i := 0; i < len(concat); i++ {
		cur.Next = concat[i]
		cur = cur.Next
	}
	cur.Next = nil
	return dump.Next
}

// classic solution: inplace
// ./idea.png
// time		O(n)
// space	O(1)
func oddEvenList1(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}
	result := head
	odd := head
	dump := &ListNode{0, head}
	even := dump
	var oddTail *ListNode
	for odd != nil {
		temp := odd.Next
		if temp != nil {
			odd.Next = odd.Next.Next
		}
		even.Next = temp
		if odd.Next == nil {
			oddTail = odd
		}
		odd = odd.Next
		even = even.Next
	}
	oddTail.Next = dump.Next
	return result
}

func _printList(l *ListNode) {
	c := l
	for c != nil {
		fmt.Println(c.Val)
		c = c.Next
	}
}

func main() {
	a := &ListNode{1, nil}
	b := &ListNode{2, nil}
	c := &ListNode{3, nil}
	d := &ListNode{4, nil}
	e := &ListNode{5, nil}
	// f := &ListNode{6, nil}
	a.Next = b
	b.Next = c
	c.Next = d
	d.Next = e
	// e.Next = f
	_printList(oddEvenList1(a))
}

File: leetcode/328-odd-even-linked-list/main.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

"""
    navie solution
	- put the items into even and odd arrays
	- concat them together

	time		O(n)
	space	O(n)
	beats 100% (LOL the speed is the same with the classic approach)
"""


class Solution(object):
    def oddEvenList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        oddList = ListNode(0)
        evenList = ListNode(0)
        cur = head
        oddCur = oddList
        evenCur = evenList
        i = 1
        while cur != None:
            if i % 2 == 0:
                evenCur.next = ListNode(cur.val)
                evenCur = evenCur.next
            else:
                oddCur.next = ListNode(cur.val)
                oddCur = oddCur.next
            cur = cur.next
            i += 1
        oddCur.next = evenList.next
        return oddList.next


"""
    navie solution
	- put the items into even and odd arrays
	- concat them together

	time		O(n)
	space	O(n)
	beats 100% (LOL the speed is the same with the classic approach)
"""


class Solution(object):
    def oddEvenList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        oddList = ListNode(0)
        evenList = ListNode(0)
        cur = head
        oddCur = oddList
        evenCur = evenList
        i = 1
        while cur != None:
            if i % 2 == 0:
                evenCur.next = cur
                evenCur = evenCur.next
            else:
                oddCur.next = cur
                oddCur = oddCur.next
            cur = cur.next
            i += 1
        oddCur.next = evenList.next
        evenCur.next = None
        return oddList.next
